experimental/ActionPolyRing/src/Content.jl

```diff
@@ -0,0 +1,1373 @@
+   return T
+end
+
+function Base.eltype(::Type{DifferentialPolyTerms{T}}) where T <: ActionPolyRingElem{S} where S <: RingElement
+   return T
+end
+
+###############################################################################
+#
+#   Promotion rules
+#
+###############################################################################
+
+AbstractAlgebra.promote_rule(::Type{DifferencePolyRingElem{T}}, ::Type{DifferencePolyRingElem{T}}) where {T <: RingElement} = DifferencePolyRingElem{T}
+
+function AbstractAlgebra.promote_rule(::Type{DifferencePolyRingElem{T}}, ::Type{V}) where {T <: RingElement, V <: RingElement}
+   AbstractAlgebra.promote_rule(T, V) == T ? DifferencePolyRingElem{T} : Union{}
+end
+
+AbstractAlgebra.promote_rule(::Type{DifferentialPolyRingElem{T}}, ::Type{DifferentialPolyRingElem{T}}) where {T <: RingElement} = DifferentialPolyRingElem{T}
+
+function AbstractAlgebra.promote_rule(::Type{DifferentialPolyRingElem{T}}, ::Type{V}) where {T <: RingElement, V <: RingElement}
+   AbstractAlgebra.promote_rule(T, V) == T ? DifferentialPolyRingElem{T} : Union{}
+end
+
+###############################################################################
+#
+#   Random generation
+#
+###############################################################################
+
+rand(rng::AbstractRNG, apr::ActionPolyRing, term_range::AbstractUnitRange{Int},
+     exp_bound::AbstractUnitRange{Int}, v...) = apr(rand(rng, __upr(apr), term_range, exp_bound, v...))
+
+rand(apr::ActionPolyRing, term_range, exp_bound, v...) = rand(Random.default_rng(), apr, term_range, exp_bound, v...)
+
+
+###############################################################################
+#
+#   Conformance test element generation
+#
+###############################################################################
+
+function ConformanceTests.generate_element(R::ActionPolyRing{ZZRingElem})
+    return rand(R, 0:4, 0:10, -10:10)
+end
+
+function ConformanceTests.generate_element(R::ActionPolyRing{ZZModRingElem})
+    return rand(R, 0:4, 0:10, -10:10)
+end
+
+###############################################################################
+#
+#  Misc 
+#
+###############################################################################
+
+symbols(apr::ActionPolyRing) = symbols(__upr(apr))[__perm_for_sort(apr)]
+
+canonical_unit(apre::ActionPolyRingElem) = canonical_unit(data(apre))
+
+###############################################################################
+#
+#  Update polynomials 
+#
+###############################################################################
+
+# The multivariate polynomial ring parent(data(data(dpre))) will always equal what has been parent(dpre).upoly_ring.mpoly_ring at the time of its creation, so
+# we need a way update this, e.g. after adding variables. This is mandatory for some methods to work proper and also just good practise.
+
+function __update_internals!(dpre::DifferencePolyRingElem)
```

Comment:
these functions heavily rely on UnivPolyRing internals and shouldn't be needed. But at the time of writing this, I haven't seen any place where they are actually called, so I'll come back to this comment here